File: components/autofill_assistant/browser/android/assistant_overlay_delegate.cc
// Copyright 2018 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/autofill_assistant/browser/android/assistant_overlay_delegate.h"

#include "components/autofill_assistant/android/jni_headers/AssistantOverlayDelegate_jni.h"
#include "components/autofill_assistant/browser/android/ui_controller_android.h"

using base::android::AttachCurrentThread;

namespace autofill_assistant {

AssistantOverlayDelegate::AssistantOverlayDelegate(
    UiControllerAndroid* ui_controller)
    : ui_controller_(ui_controller) {
  java_assistant_overlay_delegate_ = Java_AssistantOverlayDelegate_create(
      AttachCurrentThread(), reinterpret_cast<intptr_t>(this));
}

AssistantOverlayDelegate::~AssistantOverlayDelegate() {
  Java_AssistantOverlayDelegate_clearNativePtr(
      AttachCurrentThread(), java_assistant_overlay_delegate_);
}

void AssistantOverlayDelegate::OnUnexpectedTaps(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& jcaller) {
  ui_controller_->OnUnexpectedTaps();
}

base::android::ScopedJavaGlobalRef<jobject>
AssistantOverlayDelegate::GetJavaObject() {
  return java_assistant_overlay_delegate_;
}

}  // namespace autofill_assistant

File: components/autofill_assistant/android/java/src/org/chromium/components/autofill_assistant/overlay/AssistantOverlayDelegate.java
// Copyright 2018 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.autofill_assistant.overlay;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;

/** Delegate for the overlay. */
@JNINamespace("autofill_assistant")
class AssistantOverlayDelegate {
    private long mNativeAssistantOverlayDelegate;

    @CalledByNative
    private static AssistantOverlayDelegate create(long nativeAssistantOverlayDelegate) {
        return new AssistantOverlayDelegate(nativeAssistantOverlayDelegate);
    }

    private AssistantOverlayDelegate(long nativeAssistantOverlayDelegate) {
        mNativeAssistantOverlayDelegate = nativeAssistantOverlayDelegate;
    }

    /** Called after a certain number of unexpected taps. */
    void onUnexpectedTaps() {
        if (mNativeAssistantOverlayDelegate != 0) {
            AssistantOverlayDelegateJni.get().onUnexpectedTaps(
                    mNativeAssistantOverlayDelegate, AssistantOverlayDelegate.this);
        }
    }

    @CalledByNative
    private void clearNativePtr() {
        mNativeAssistantOverlayDelegate = 0;
    }

    @NativeMethods
    interface Natives {
        void onUnexpectedTaps(long nativeAssistantOverlayDelegate, AssistantOverlayDelegate caller);
    }
}
